### DISCUSSION TITLE
Collector no longer has `preprocess_fn`

### DISCUSSION THREAD

**Discussion by destin-v** [thu-ml/tianshou#1203] [Q&A] [answered]
In the latest update to Tianshou 1.1.0 it looks like the `preprocess_fn` from `Collector` has been removed.  I was using it to record full episode rollouts and running some analysis on them.  Now that `preprocess_fn` has been removed, is there a better way to get the obs, rew, terminated, truncated, info?

**Comment by Trinkle23897** [comment, marked as answer]
I think after #1196 you can set `on_episode_done_hook` or `on_step_hook` in `Collector` to achieve this.

**Comment by MischaPanch** [comment]
Hi @destin-v . It was removed a long time ago, I think (probably as part of 1.0.0) since it was unused in the codebase and increased the  complexity of the already very complicated Collector. 

Just now we have finished work on extending the Collector with hooks, it's already merged on master. I'm about to merge some internal improvements in #1214 , which however don't affect the public interfaces.

There's now a dedicated mechanism for processing batches at step and at episode completion by passing hooks to the collector. Have a look at the docstrings and examples there pls. We will add more documentation and tutorials on these new features with the next release.

So for your experiments, I suggest you move to the version currently on master

**Reply by MischaPanch** [reply]
As a benefit, the internals of Collector are now much more documented

**Comment by destin-v** [comment]
Okay, sounds good.  I would recommend that the [Cheat Sheet](https://tianshou.org/en/stable/01_tutorials/07_cheatsheet.html) remove the documentation for `preprocess_fn` since it does not exist anymore.  We will likely want some documentation on `on_episode_done_hook` and `on_step_hook`.

**Reply by MischaPanch** [reply]
Documentation is generally severely lacking.. I am planning to devote a lot of time for it in September

**Reply by MischaPanch** [reply]
Feel free to start on that if you want to and have time! If that's the case, let's have a short call to coordinate
